icicle/appUtils/msm/msm.cu

```diff
@@ -395,12 +396,14 @@ void bucket_method_msm(
   NUM_THREADS = min(1 << 8, nof_runs);
   NUM_BLOCKS = (nof_runs + NUM_THREADS - 1) / NUM_THREADS;
   add_ones_kernel<<<NUM_BLOCKS, NUM_THREADS, 0, stream>>>(d_points, d_scalars, ones_results, size, run_length);
+  CHECK_CUDA_ERROR();
```

Comment:
`CHECK_CUDA_ERROR` should be used to wrap CUDA methods

`CHECK_CUDA_ERROR(<cuda_method>)`

I think u want to use 

`CHECK_LAST_CUDA_ERROR()`

This mehod calls `cudaGetLastError` [docs here](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__ERROR.html#group__CUDART__ERROR_1g3529f94cb530a83a76613616782bd233)

Reply:
CHECK_SYNC_DEVICE_ERROR on the other hand is blocking and will wait for a computation to complete before returning error.

Reply:
I followed the best practice suggestion from [here](https://leimao.github.io/blog/Proper-CUDA-Error-Checking/):

**Check CUDA asynchronous error in development by synchronization and error checking after kernel launch calls and disable it in production.**

I have a `#DEFINE ERROR_MODE` in `error_handler.cuh` that selects either TEST mode (development) or PRODUCTION:

1. In TEST mode we want to localize errors even at expense of the speed. This is why I'd like to use CHECK_SYNC_DEVICE_ERROR.

In PRODUCTION mode we prioritize the speed over error localization, so  CHECK_LAST_CUDA_ERROR

Does this make sense? If yes, we can figure out how exactly we select between the TEST and PRODUCTION.

Reply:
There is another point, that I'd like put into separate comment. CUDA (and the article above) differentiate between recoverable (i.e. not enough memory) and non-recoverable (null pointer).

Do we want to take care of such errors? Personally, I'd just kill the process. K8s would take care of resilience by detecting the kill and restarting the app. For mining this shouldn't be a problem. Not sure about other apps.

Reply:
@svpolonsky what u did is ok

Im just saying ur using the wrong method, if you look at [CHECK_CUDA_ERROR](https://github.com/ingonyama-zk/icicle/blob/2387086f4376372eb4f86b683697e094773b2d7d/icicle/utils/error_handler.cuh#L4) it requires a parameter input.

Reply:
@svpolonsky use CHECK_LAST_CUDA_ERROR

Reply:
@ImmanuelSegol Stas is trying to use the defined names that were added [here](https://github.com/ingonyama-zk/icicle/pull/259/files#diff-4e31376046899777077b6965bf464226629e39527466bd8b85edf164988710e5R45-R48) so calling it depends on whether it is TEST or PRODUCTION.

Though @svpolonsky, I agree with Immanuel that shadowing the already defined [CHECK_CUDA_ERROR](https://github.com/ingonyama-zk/icicle/blob/2387086f4376372eb4f86b683697e094773b2d7d/icicle/utils/error_handler.cuh#L4) might not be the best. Can we change the TEST/PROD name to something else?